### DISCUSSION TITLE
Form表单中无法感知整个表单的必填项是否都已填写，从而控制提交按钮的可用状态

### DISCUSSION THREAD

**Discussion by leegsen7** [ant-design/ant-design-mobile#5237] [Q&A] [answered]
希望的场景交互是必填项都已填写，提交按钮才是可用的，点击提交按钮，再走rules校验。

**Comment by damonyoungcc** [comment]
我觉得使用`useWatch`可以解决，[戳这个例子](https://mobile.ant.design/zh/components/form#usewatch)

**Reply by leegsen7** [reply]
我之前有看过这个文档例子。使用`useWatch`来解决我描述的问题无疑也是繁琐的。现在Form表单中是无法感知整个表单的必填项是否都已填写，也是通过额外的状态处理才能去感知。就如这个`useWatch`的例子，如果你要控制按钮（当必填项都已填写）可用的话，那就得额外写一套判断逻辑了，而现在view层也是有一套判断逻辑，相当于要维护两套类似的逻辑。我认为更好的方式是由Form去感知收集Form.Item，新增一个Form回调函数属性onCheckRequired来传递Form表单必填项的填写情况

**Reply by awmleer** [reply]
可以用 `form. getFieldsError` 么？我其实没试过

要不写个 codesandbox 上的 demo 吧，我们都在 demo 的基础上改一改试试

**Reply by damonyoungcc** [reply]
是的，老哥你要不写一个demo，说说想要实现什么具体的功能，大家一起改改试试，肯定有比较优雅的方案的。

**Reply by leegsen7** [reply]
> 希望的场景交互是必填项都已填写，提交按钮才是可用的，点击提交按钮，再走rules校验。

@yangdepp @awmleer 

实现这个交互的话，用已有的`useWatch`例子来演示下，代码组织很费劲。

[https://codesandbox.io/s/compassionate-stitch-9xor8n?file=/src/App.js](https://codesandbox.io/s/compassionate-stitch-9xor8n?file=/src/App.js)

我建议一种新方案，由Form去感知收集name，新增一个Form回调函数属性`onCheckRequired`来传递Form表单必填项的填写情况
```jsx
<Form
  form={form}
  onCheckRequired={flag => setCheckRequired(flag)}
/>
```

**Comment by awmleer** [comment, marked as answer]
找到了 antd 的 demo： https://codesandbox.io/s/jjps3v?file=/demo.js

可以先参考着写一下，不过它这个也实现的不是很好，各种粗暴的强制更新

**Reply by leegsen7** [reply]
确实，都是很硬的写法。
